programs/test/metatest.c

```diff
@@ -0,0 +1,273 @@
+/** \file metatest.c
+ *
+ *  \brief Test features of the test framework.
+ */
+
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+ */
+
+#define MBEDTLS_ALLOW_PRIVATE_ACCESS
+
+#include <mbedtls/platform.h>
+#include <mbedtls/platform_util.h>
+#include "test/helpers.h"
+#include "test/macros.h"
+
+#include <stdio.h>
+#include <string.h>
+
+#if defined(MBEDTLS_THREADING_C)
+#include <mbedtls/threading.h>
+#endif
+
+
+/* This is an external variable, so the compiler doesn't know that we're never
+ * changing its value.
+ *
+ * TODO: LTO (link-time-optimization) would defeat this.
+ */
+int false_but_the_compiler_does_not_know = 0;
+
+
+/****************************************************************/
+/* Test framework features */
+/****************************************************************/
+
+void meta_test_fail(const char *name)
+{
+    (void) name;
+    mbedtls_test_fail("Forced test failure", __LINE__, __FILE__);
+}
+
+
+/****************************************************************/
+/* Platform features */
+/****************************************************************/
+
+void null_pointer_dereference(const char *name)
+{
+    (void) name;
+    volatile char *p;
+    mbedtls_platform_zeroize((void *) &p, sizeof(p));
+    mbedtls_printf("%p -> %u\n", p, (unsigned) *p);
+}
+
+void null_pointer_call(const char *name)
+{
+    (void) name;
+    unsigned (*p)(void);
+    mbedtls_platform_zeroize(&p, sizeof(p));
+    /* The pointer representation may be truncated, but we don't care:
+     * the only point of printing it is to have some use of the pointer
+     * to dissuade the compiler from optimizing it away. */
+    mbedtls_printf("%lx() -> %u\n", (unsigned long) (uintptr_t) p, p());
+}
+
+
+/****************************************************************/
+/* Memory */
+/****************************************************************/
+
+void read_after_free(const char *name)
+{
+    (void) name;
+    volatile char *p = mbedtls_calloc(1, 1);
+    *p = 'a';
+    mbedtls_free((void *) p);
+    mbedtls_printf("%u\n", (unsigned) *p);
+}
+
+void double_free(const char *name)
+{
+    (void) name;
+    volatile char *p = mbedtls_calloc(1, 1);
+    *p = 'a';
+    mbedtls_free((void *) p);
+    mbedtls_free((void *) p);
+}
+
+void read_uninitialized_stack(const char *name)
+{
+    (void) name;
+    volatile char buf[1];
```

Comment:
Why is `buf` volatile here? Could that stop MSAN from treating the read as uninitialised?

Reply:
`buf` is volatile so that the compiler treats `*buf` as an access on the stack, and not as a read from a register.

If that stops MSan from seeing the read as uninitialized (but why would it?), the CI will fail.

Reply:
It wouldn't be unreasonable for MSan to allow a read of an uninitialised volatile (because something unknown to MSan might have initialised it). Whether it can or does do that I don't know.

`char buf[1]; char * volatile p = buf;` and then reading `*p` would cover this.

Reply:
We know MSan doesn't do this with current versions of the tooling because the test passes. I don't know why it would start doing that. But anyway making `buf` non-volatile and inserting a volatile `p` also seems to work to prevent a compiler optimization, so ok, I'll put that in.